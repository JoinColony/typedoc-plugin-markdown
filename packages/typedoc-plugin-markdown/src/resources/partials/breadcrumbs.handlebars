{{~#if @hasReadmeFile}}

[Readme]({{{relativeURL @options.entryDocument}}}) \| {{ ' ' }}[{{@globalsName}}]({{{relativeURL @project.url}}})

{{~/if~}}

{{~#reflectionTree this~}}

{{#each tree}} {{#if @first}} / {{/if}} [{{name}}]({{{urlTo .}}}) {{#unless @last}} / {{/unless}} {{/each}} 

{{~/reflectionTree~}}