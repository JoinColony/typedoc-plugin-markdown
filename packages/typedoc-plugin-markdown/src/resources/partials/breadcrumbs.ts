import { DeclarationReflection, PageEvent, ProjectReflection } from 'typedoc';
import { MarkdownThemeContext } from '../../theme-context';

import { escapeChars } from '../../utils/format';

export const breadcrumbsPartial = (
  context: MarkdownThemeContext,
  props: PageEvent<DeclarationReflection | ProjectReflection>,
) => {
  const { entryPoints, entryDocument, readme } = context.options;

  function breadcrumbs() {
    const project = context.project();
    if (!project) {
      return '';
    }
    const hasReadmeFile = !readme.endsWith('none');
    const breadcrumbs: string[] = [];
    const globalsName = entryPoints.length > 1 ? 'Modules' : 'Exports';
    breadcrumbs.push(
      props.url === entryDocument
        ? project.name
        : `[${project.name}](${context.relativeURL(entryDocument)})`,
    );
    if (hasReadmeFile) {
      breadcrumbs.push(
        props.url === project.url
          ? globalsName
          : `[${globalsName}](${context.relativeURL('modules.md')})`,
      );
    }
    const breadcrumbsOut = breadcrumb(props, props.model, breadcrumbs);
    return breadcrumbsOut;
  }

  function breadcrumb(page: PageEvent, model: any, md: string[]) {
    if (model && model.parent) {
      breadcrumb(page, model.parent, md);
      if (model.url) {
        md.push(
          page.url === model.url
            ? `${escapeChars(model.name)}`
            : `[${escapeChars(model.name)}](${context.urlTo(model)})`,
        );
      }
    }
    return md.join(' / ');
  }

  return breadcrumbs();
};
