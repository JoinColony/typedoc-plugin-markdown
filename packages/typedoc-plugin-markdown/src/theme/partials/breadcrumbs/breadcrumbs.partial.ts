import { DeclarationReflection, PageEvent, ProjectReflection } from 'typedoc';
import { MarkdownThemeRenderContext } from '../../theme.context';
import { escapeChars } from '../../theme.utils';

export const breadcrumbsPartial = (
  context: MarkdownThemeRenderContext,
  props: PageEvent<ProjectReflection | DeclarationReflection>,
) => {
  const entryPoints = context.getOption('entryPoints');
  const entryDocument = context.getOption('entryDocument');
  const readme = context.getOption('readme');

  const md: string[] = [];

  const globalsName = entryPoints.length > 1 ? 'Modules' : 'Exports';
  md.push(
    props.url === entryDocument
      ? context.project?.name
      : `[${context.project?.name}](${context.relativeURL(entryDocument)})`,
  );
  if (!readme.endsWith('none')) {
    md.push(
      props.url === context.project?.url
        ? globalsName
        : `[${globalsName}](${context.relativeURL(context.modulesFileName)})`,
    );
  }

  const breadcrumbsOut = breadcrumb(context, props, props.model, md);
  return breadcrumbsOut;
};

const breadcrumb = (
  context: MarkdownThemeRenderContext,
  page: PageEvent,
  model: any,
  md: string[],
) => {
  if (model && model.parent) {
    breadcrumb(context, page, model.parent, md);
    if (model.url) {
      md.push(
        page.url === model.url
          ? `${escapeChars(model.name)}`
          : `[${escapeChars(model.name)}](${context.relativeURL(model.url)})`,
      );
    }
  }
  return md.join(' / ');
};
