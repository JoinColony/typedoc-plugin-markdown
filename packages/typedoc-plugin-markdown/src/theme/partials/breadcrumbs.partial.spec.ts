import { ProjectReflection } from 'typedoc';
import { breadcrumbsPartial } from './breadcrumbs.partial';

describe(`Breadcrumbs Partial:`, () => {
  describe(`(with readme)`, () => {
    let project: ProjectReflection;

    beforeEach(async () => {
      project = {
        name: 'project-name',
        url: 'modules.md',
      } as ProjectReflection;
    });
    test(`should compile README breadcrumbs'`, () => {
      expect(
        breadcrumbsPartial(global.getMockContext(), {
          project: project,
          model: project,
          url: 'README.md',
        } as any),
      ).toMatchSnapshot();
    });

    test(`should compile module breadcrumbs'`, () => {
      expect(
        breadcrumbsPartial(global.getMockContext(), {
          project: project,
          model: {
            parent: { url: 'modules.md' },
            url: 'breadcrumbs.md',
            name: 'breadcrumbs',
          },
          url: 'breadcrumbs.md',
        } as any),
      ).toMatchSnapshot();
    });

    test(`should compile class breadcrumbs'`, () => {
      expect(
        breadcrumbsPartial(global.getMockContext(), {
          project: project,
          model: {
            parent: {
              parent: { url: 'modules.md' },
              url: 'breadcrumbs.md',
              name: 'breadcrumbs.md',
            },
            url: 'breadcrumbs.Breadcrumbs.md',
            name: 'Breadcrumbs',
          },
          url: 'breadcrumbs.Breadcrumbs.md',
        } as any),
      ).toMatchSnapshot();
    });
  });

  describe(`(without readme)`, () => {
    let project: ProjectReflection;

    beforeEach(async () => {
      project = {
        name: 'project-name',
        url: 'modules.md',
      } as ProjectReflection;
    });
    test(`should compile README breadcrumbs'`, () => {
      expect(
        breadcrumbsPartial(global.getMockContext({ readme: 'none' }), {
          project: project,
          model: project,
          url: 'README.md',
        } as any),
      ).toMatchSnapshot();
    });

    test(`should compile module breadcrumbs'`, () => {
      expect(
        breadcrumbsPartial(global.getMockContext({ readme: 'none' }), {
          project: project,
          model: {
            parent: { url: 'modules.md' },
            url: 'breadcrumbs.md',
            name: 'breadcrumbs',
          },
          url: 'breadcrumbs.md',
        } as any),
      ).toMatchSnapshot();
    });

    test(`should compile class breadcrumbs'`, () => {
      expect(
        breadcrumbsPartial(global.getMockContext({ readme: 'none' }), {
          project: project,
          model: {
            parent: {
              parent: { url: 'modules.md' },
              url: 'breadcrumbs.md',
              name: 'breadcrumbs.md',
            },
            url: 'breadcrumbs.Breadcrumbs.md',
            name: 'Breadcrumbs',
          },
          url: 'breadcrumbs.Breadcrumbs.md',
        } as any),
      ).toMatchSnapshot();
    });
  });
});
