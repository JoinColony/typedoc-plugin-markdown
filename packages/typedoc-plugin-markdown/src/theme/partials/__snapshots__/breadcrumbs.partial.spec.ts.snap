// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs Partial: (with readme) should compile README breadcrumbs' 1`] = `" / [Exports](( relativeURL ))"`;

exports[`Breadcrumbs Partial: (with readme) should compile class breadcrumbs' 1`] = `"[undefined](( relativeURL )) / [Exports](( relativeURL )) / [breadcrumbs.md](( relativeURL )) / Breadcrumbs"`;

exports[`Breadcrumbs Partial: (with readme) should compile module breadcrumbs' 1`] = `"[undefined](( relativeURL )) / [Exports](( relativeURL )) / breadcrumbs"`;

exports[`Breadcrumbs Partial: (without readme) should compile README breadcrumbs' 1`] = `""`;

exports[`Breadcrumbs Partial: (without readme) should compile class breadcrumbs' 1`] = `"[undefined](( relativeURL )) / [breadcrumbs.md](( relativeURL )) / Breadcrumbs"`;

exports[`Breadcrumbs Partial: (without readme) should compile module breadcrumbs' 1`] = `"[undefined](( relativeURL )) / breadcrumbs"`;
